<?xml version="1.0"?>

<launch>

<!-- Argument names for amcl parameters-->
  <arg name="use_map_topic"   default="true"/>
  <arg name="initial_pose_x"  default="0.0"/>
  <arg name="initial_pose_y"  default="0.0"/>
  <arg name="initial_pose_a"  default="0.0"/>
  <arg name="odom_frame_id"   default="odom"/>
  <arg name="base_frame_id"   default="base_frame"/>
  <arg name="global_frame_id" default="map"/>

<!-- Argument names for hector parameters-->
  <arg name="tf_map_scanmatch_transform_frame_name" default="scanmatcher_frame"/>
  <arg name="base_frame" default="base_frame"/>
  <arg name="odom_frame" default="odom_frame"/>
  <arg name="pub_map_odom_transform" default="false"/>
  <arg name="scan_subscriber_queue_size" default="5"/>
  <arg name="scan_topic" default="updatedScan"/>
  <arg name="map_size" default="2048"/>

  <arg name="map_name" default="living3.yaml"/>
  <arg name="speed" default="3.0"/>

<!-- Uncomment in case of rosbag -->
  <!--<param name="/use_sim_time" value="true"/>-->


<!-- Map server -->
  <node pkg="map_server" type="map_server" name="map_server" args="$(find f1tenth_2dnav)/maps/$(arg map_name)"  output="screen" />

<!-- Hector node -->
  <node name="hector_mapping" pkg="hector_mapping" type="hector_mapping" output="screen">
      <remap from="map" to="mapcurrent" />
      <param name="map_frame" value="map" />
      <param name="base_frame" value="base_frame" />

      <param name="odom_frame" value="scanmatch_odom" />
    
      <!-- Tf use -->
      <param name="use_tf_scan_transformation" value="true"/>
      <param name="use_tf_pose_start_estimate" value="false"/>
      <param name="pub_map_odom_transform" value="$(arg pub_map_odom_transform)"/>

      <param name="pub_odometry" value="true"/>

    
      <!-- Map size / start point -->
      <param name="map_resolution" value="0.050"/>
      <param name="map_size" value="$(arg map_size)"/>
      <param name="map_start_x" value="0.5"/>
      <param name="map_start_y" value="0.5" />
      <param name="map_multi_res_levels" value="2" />
    
      <param name="update_factor_free" value="0.4" />
      <param name="update_factor_occupied" value="0.9" />    
      <param name="map_update_distance_thresh" value="0.01" />
      <param name="map_update_angle_thresh" value="0.01" />
      <param name="laser_z_min_value" value = "-1.0" />
      <param name="laser_z_max_value" value = "1.0" />

      <param name="advertise_map_service" value="false" />
    
      <param name="scan_subscriber_queue_size" value="5"/>
      <param name="scan_topic" value="$(arg scan_topic)"/>
      <param name="pub_map_scanmatch_transform" value="false" />
      <param name="tf_map_scanmatch_transform_frame_name" value="scanmatcher_frame" />
  </node>

<!-- Packages for odom from hector and amcl-->
  <node pkg="f1tenth_2dnav" type="hectorOdom" name="hectorOdom" output="screen"/>
  <node pkg="f1tenth_2dnav" type="amclOdom" name="amclOdom" output="screen"/>

<!-- Amcl node -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <!-- Publish scans from best pose at a max of 10 Hz -->
    <param name="odom_model_type" value="diff"/>
    <param name="transform_tolerance" value="0.2" />
    <param name="gui_publish_rate" value="100.0"/>
    <param name="laser_max_beams" value="90"/>
    <param name="min_particles" value="500"/>
    <param name="max_particles" value="5000"/>
    <param name="kld_err" value="0.05"/>
    <param name="kld_z" value="0.99"/>
    <param name="resample_interval" value="1"/>

    <param name="save_pose_rate" value="100.0"/>

    <param name="odom_alpha1" value="0.2"/>
    <param name="odom_alpha2" value="0.2"/>
    <!-- translation std dev, m -->
    <param name="odom_alpha3" value="0.8"/>
    <param name="odom_alpha4" value="0.2"/>
    <param name="laser_z_hit" value="0.5"/>
    <param name="laser_z_short" value="0.05"/>
    <param name="laser_z_max" value="0.05"/>
    <param name="laser_z_rand" value="0.5"/>
    <param name="laser_sigma_hit" value="0.2"/>
    <param name="laser_lambda_short" value="0.1"/>
    <param name="laser_lambda_short" value="0.1"/>
    <param name="laser_model_type" value="likelihood_field"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="update_min_d" value="0.1"/>
    <param name="update_min_a" value="0.5"/>
    <param name="odom_frame_id" value="$(arg odom_frame_id)"/>
    <param name="base_frame_id" value="$(arg base_frame_id)"/> 
    <param name="global_frame_id" value="$(arg global_frame_id)"/>
    <param name="resample_interval" value="1"/>
    <param name="transform_tolerance" value="0.1"/>
    <param name="recovery_alpha_slow" value="0.0"/>
    <param name="recovery_alpha_fast" value="0.0"/>
    <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
    <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
    <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
    <remap from="scan"                      to="$(arg scan_topic)"/>

    <param name="laser_likelihood_max_dist" value = "6.0"/>
  </node>

<!-- move_base node -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <!-- Internal parameters of move_base -->
   
    <param name="controller_frequency" value="15"/>
    <!--<param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />-->

    <!-- External parameters of move_base -->

    <rosparam file="$(find f1tenth_2dnav)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find f1tenth_2dnav)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find f1tenth_2dnav)/config/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find f1tenth_2dnav)/config/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find f1tenth_2dnav)/config/base_local_planner_params.yaml" command="load" />
  </node>
  
  <node pkg="f1tenth_2dnav" type="cmd_vel_to_drive_param.py" name="cmd_vel_to_drive_param" output="screen">
    <param name="speed" value="$(arg speed)"/>
  </node>
</launch>
